### DISCUSSION TITLE
Advertised vs Non-Advertised Shortcuts in WiX v4  Designer Strategy

### DISCUSSION THREAD

**Discussion by chrpai** [wixtoolset/issues#7545] [Questions]
This is an open ended question with a specific question at the end.

A couple of decades ago I developed a bad taste for advertised shortcuts.  They weren't the source of the problem but rather a symptom.  The root causes are mostly gone these days and I'm reevaluating my thoughts and would like feedback.  

I've been using merge modules for distributed setup development.  In this context non-adveristised shortcuts don't give the annoying ICE errors as it's using SpecialFolder.GUID instead of SpecialFolder and the ICEs don't pick up on this.  Moving to fragment/wixlib the ICE errors start occuring.   Two ways around this seem to be author the directory table and use setproperty custom actions to mimic the way a merge module shortcut would work or use an advertised shortcut and simply set DISABLEADVTSHORTCUTS to 1.  
 
Another approach would be to embrace advertised shortcuts. Is there a way to tell WiX to simply extract an embedded icon resource from the target keyfile?  Other tools I've used in the past had something like this but so far I've had to modify my upstream build to make an .ico available and manually author it into the fragment.

**Comment by chrpai** [comment]
Any thoughts from anyone?

**Comment by chrpai** [comment]
@robmen @barnson  if anyone feels like helping.  I recently wrote an issue on anyonymous directories and ICE 64 and what I'm really looking for is simple shortcut authoring best practices and avoid things like ICEs.

**Reply by barnson** [reply]
There could be a syntactic sugar feature to simplify everything.

**Comment by chrpai** [comment]
@barnson @robmen My goal here was to modernize IsWiX to v4 standards migrating from merge modules to fragments in the process.  I'll do it again when v5 comes out.  I had engaged Sean for a couple hours to get his review and feedback on what v4 authoring should look like.  I'd be willing to do this again with you or Rob. 

The specific ask is to understand what authoring business logic is best compatible with v4 and easy to implement programatically. I had asked this question last summer and when I didn't get a response I just decided to put DISABLEADVTSHORTCUTS into my project template because of the friction of getting an icon file to point to.  Then I discovered the ICE64 issue using subdirectories and I'm trying to think of a way to solve for this or just disable ICE64 in my project template.  I agree ICE64 is probably FOS but I really dislike disabling them as I don't want it to look like I'm hiding anything.

**Comment by chrpai** [comment]
@bob thanks for the weekly meeting notes.  With regard to my ICE64 issue, your summary was far more useful then what was actually said in the meeting. (Ya, WiX doesn't create components down the anonymous chain because well that's just crazy)  

Your suggestion to add a RemoveFolder element works if you only have one level in the Subdirectory attribute.  Consider the folllowing fragment.  If  I omit RemoveFolder elements for 1 and 1\2  I get an error from ICE64 because only 3 is in the table.  However if I add them I still get an error from ICE64  but I see 1 2 and 3 in the table.  However the Directory table now has 5 entires.  1 and 2 are duplicated with different Ids  and 3 exists only once.

```
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <?define SourceDir="..\Deploy"?>
  <Fragment>
    <ComponentGroup Id="WebAPIFiles">
      <Component Directory="INSTALLLOCATION">
        <File Source="$(var.SourceDir)\app.exe" KeyPath="yes">
          <Shortcut Id="sc1" Name="App" Directory="DesktopFolder" Subdirectory="1\2\3" Advertise="yes" />
        </File>
        <RemoveFolder Id="rf1" Directory="DesktopFolder" Subdirectory="1" On="uninstall" />
        <RemoveFolder Id="rf2" Directory="DesktopFolder" Subdirectory="1\2" On="uninstall" />
        <RemoveFolder Id="rf3" Directory="DesktopFolder" Subdirectory="1\2\3" On="uninstall" />
      </Component>
    </ComponentGroup>
  </Fragment>
</Wix>
```

**Comment by chrpai** [comment]
I created #7861 to track this.

**Comment by chrpai** [comment]
For completeness, this actually works.  It's just a lot more verbose and takes a lot more logic to read and write.

```
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <?define SourceDir="..\Deploy"?>
  <Fragment>
    <ComponentGroup Id="WebAPIFiles">
      <Component Directory="INSTALLLOCATION">
        <File Source="$(var.SourceDir)\website\api.exe" KeyPath="yes">
          <Shortcut Id="sc1" Name="App" Directory="Test2" Advertise="yes" />
        </File>
        <RemoveFolder Directory="Test1" On="uninstall" />
        <RemoveFolder Directory="Test2" On="uninstall" />
      </Component>
    </ComponentGroup>
    <Directory Id="DesktopFolder">
      <Directory Id="Test1" Name="1">
        <Directory Id="Test2" Name="2"></Directory>
      </Directory>
    </Directory>
  </Fragment>
</Wix>
```
